File: app/src/main/java/com/github/ilikeyourhat/sudokusolver/solving/deduction/algorithm/HiddenSinglesAlgorithm.java
package com.github.ilikeyourhat.sudokusolver.solving.deduction.algorithm;

import com.github.ilikeyourhat.sudokusolver.model.Field;
import com.github.ilikeyourhat.sudokusolver.model.Region;
import com.github.ilikeyourhat.sudokusolver.model.hint.SudokuHintGrid;

import java.util.*;

public class HiddenSinglesAlgorithm extends DeductionAlgorithm {

    public static class Factory implements DeductionAlgorithm.Factory {

        public HiddenSinglesAlgorithm instance(List<Region> regions, SudokuHintGrid sudokuHintGrid) {
            return new HiddenSinglesAlgorithm(regions, sudokuHintGrid);
        }

    }

    private HiddenSinglesAlgorithm(List<Region> regions, SudokuHintGrid possibilities) {
        super(regions, possibilities);
    }

    @Override
    protected void solve(Region region) {
        for (int value = 1; value <= region.size(); value++) {
            Set<Field> occurrences = new HashSet<>();
            for (Field field : region) {
                if (possibilities.contains(field, value)) {
                    occurrences.add(field);
                }
            }
            if (occurrences.size() == 1) {
                Field field = (Field) occurrences.toArray()[0];
                field.set(value);
                notifyChange();
            }
        }
    }

}

File: app/src/main/kotlin/com/github/ilikeyourhat/kudoku/solving/deduction/algorithm/HiddenSinglesAlgorithm.kt
package com.github.ilikeyourhat.kudoku.solving.deduction.algorithm

import com.github.ilikeyourhat.kudoku.model.Field
import com.github.ilikeyourhat.kudoku.model.Region
import com.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid

class HiddenSinglesAlgorithm(
    regions: List<Region>,
    possibilities: SudokuHintGrid
) : DeductionAlgorithm(regions, possibilities) {

    class Factory : DeductionAlgorithm.Factory {
        override fun instance(regions: List<Region>, possibilities: SudokuHintGrid): HiddenSinglesAlgorithm {
            return HiddenSinglesAlgorithm(regions, possibilities)
        }
    }

    override fun solve(region: Region) : Boolean {
        for (value in 1..region.size()) {
            val occurrences = mutableSetOf<Field>()
            for (field in region) {
                if (possibilities.contains(field, value)) {
                    occurrences.add(field)
                }
            }
            if (occurrences.size == 1) {
                occurrences.single().set(value)
                return true
            }
        }
        return false
    }
}
